### DISCUSSION TITLE
custom modifiers

### DISCUSSION THREAD

**Discussion by lukasheinrich** [scikit-hep/pyhf#1627] [Ideas] [answered]
continuing some discussion around #1625 and #1188 and #850

with those changes it's possible to have a spec: 

```
m = pyhf.Model(
    {
        'channels': [
            {
                'name': 'singlechannel',
                'samples': [
                    {
                        'name': 'signal',
                        'data': [10] * 20,
                        'modifiers': [
                            {'name': 'mu1', 'type': 'normfactor', 'data': None},
                            {'name': 'mu2', 'type': 'normfactor', 'data': None},
                            {'name': 'f_of_mus', 'type': 'customfunc', 'data': {'deps': ['mu1,mu2']}},
                        ],
                    },
                    {'name': 'background', 'data': [300] * 20, 'modifiers': []},
                ],
            }
        ]
    },
    modifier_set=modifier_set,
    poi_name='mu1',
    validate=False,
)
```

wherre `customfunc` uses code such as 

```
    def compute_function(self,pars):
        mu1,mu2  = self.param_viewer.get(pars)
        computed = pyhf.tensorlib.astensor([(mu1+mu2)**2])
        return computed
```

where `singlemod` is a multiplicative modifier to the overall rat that is a function of µ1 and µ2 such that the total rate is

```
rate = func(mu1,mu2)*mu1*mu2*signal + bkg
```

`func` in this case is only a function of **already-existing** parameters, functions that declare entirely new vaiables, but this would require some add. thinking on syntax. 

would this work @balunas @mswiatlo @alexander-held ?

from @balunas's post it seems like htere some new parameters `k1,k2,k3` need to be introduced?

**Comment by mswiatlo** [comment]
Hi @lukasheinrich ,

This is great, thanks! So in our particular case we do need parameters/constants available but I guess they could be hardcoded into func? I hope @balunas can clarify as he's worked with this more direclty

**Reply by lukasheinrich** [reply]
constants of course can be in the function but new parameters (not implied through the use of any otherr  modifiere) would need to be declared manually

**Reply by alexander-held** [reply]
Is it possible to declare normfactors that are not attached to any samples, but would be parameters in the fit and indirectly have an effect on samples through an expression depending on them? This is required for a few examples I have seen. It may be possible to work around this by attaching them to a sample that's forced to contribute zero yield via another constant normfactor acting on it, but maybe there is a more straightforward solution.

If the above works, constants could also be defined as constant normfactors (possibly convenient if multiple values of these constants need to be tested).

**Reply by lukasheinrich** [reply]
It should not be difficult to allow users to specify entirely new parameters. We might need to change a bit how we use the "patameter" section (currently it assumes it's only overriding defaults not adding anything) but it's not a fundamental constrain

**Reply by alexander-held** [reply]
The implementation looks nice to me. I guess possibly modifying the schema to support parameter creation outside of samples, or possibly inlining the function definition for easy cases, is out of scope for #1625 anyway. One small thing I noticed:
```python
'data': {'deps': ['mu1,mu2']}}
```
I assume the argument should be a list of strings, `["mu1", "mu2"]`?

**Reply by lukasheinrich** [reply]
yes that's right.. 

re scope: 1625 will add generic support for custom modifiers .. adding a newe modifier for this specific pure-function-like behavior would come afterwards

The schema alrerady has keys

```
measurements
    parameters cfg
channels
observations
version
```

we could restructure it a bit as 

```
measurements
    parameters cfg
model
    parameters cfg
    channels
observations
version
```

**Comment by lukasheinrich** [comment, marked as answer]
Hi again - 
we merged #1625 and now we can at least technically make this happen

Here is an example notebook that implements an dummy funcntio `f(k1,k2) = (k1+k2)**2` and adds that only to the signal component
https://gist.github.com/lukasheinrich/d100605d5473ea4bfa0980e4f78b01b3

there is still a lot of boilerplate associated witth this which will reduce in folow up PRs
